### PR TITLE
Email comparison view: Add the ability to present regional discounts

### PR DESCRIPTION
**Pull request by olaseni**
To support the regional pricing work we are currently wrangling in the backend, we need a few pieces in the frontend to ensure that we display discounts and product offers properly.

Sometimes the target product also has a regular sale coupon attached and we have to factor that into the display as well.

#### Proposed Changes

* Extend the introductory product offer display to support non-zero discounts. We used to only display zero pricing, optimized for free trials.
* Extend the offer eligibility for email products to include new domains in the cart. 

#### Screenshot

This is an example of how the discount is presented when both the regional offer and a sale coupon are active:
 
Screen Shot 2022-09-02 at 8 10 27 AM

#### Testing Instructions

* See setup instructions from 2d726-pb
* Load the stacked email comparison view in your local dev server i.e. http://calypso.localhost:3000/email/[:site] . You can also use [Calypso Live](https://github.com/Automattic/wp-calypso/pull/66153#issuecomment-1201968909) to navigate that view. You can also reach this view by attempting to purchase a new domain, which leads you to the email upsell. Either way, always ensure to use a domain that has not been used to purchase email before.
* The best way to test the price information unit is to have about four users, with four different currencies: 
    * MXN - introductory offer
    * PHP - introductory offer
    * TRY - introductory offer
    * USD - default system billing plans
    * You can also mix in a sale coupon ( optional )
* Scenarios to test:
    - Price display for each of the currencies is in conformance with the resolution here: pbLl1t-1iP-p2#comment-1253
    - Ensure that the offer discount is displayed as a percentage
    - Ensure that the display works well when a sale discount is also active

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

File: client/my-sites/email/email-providers-comparison/price/google-workspace.tsx
Comment (wongasy) [inline client/my-sites/email/email-providers-comparison/price/google-workspace.tsx]: To simplify this PR and focus on the test cases related to regional discounts, can we leave out the change that dynamically generates the free trial length, and handle it in a separate PR?

File: client/my-sites/email/email-providers-comparison/price/google-workspace.tsx
Comment (wongasy) [inline client/my-sites/email/email-providers-comparison/price/google-workspace.tsx]: nit, I wonder if it helps with readability if we were to prioritize `! isEligibleForIntroductoryOfferFreeTrial` as the first condition. It seems to better translate the logic into:
```
if ( isEligibleForIntroductoryOfferFreeTrial ) {
    // Render free trial
} else {
    // Render other non-free-trial offers
}
```

File: client/my-sites/email/email-providers-comparison/price-with-interval/index.tsx
Comment (wongasy) [inline client/my-sites/email/email-providers-comparison/price-with-interval/index.tsx:22]:
```suggestion
    }
	if ( isDiscounted ) {
		return product.sale_cost;
	}
	if ( isEligibleForIntroductoryOffer ) {
		return product.introductory_offer.cost_per_interval;
	}
	return null;
```

Comment (wongasy) [reply client/my-sites/email/email-providers-comparison/price-with-interval/index.tsx:22]: Wondering if it's easier to read if we were to separate out the conditions? I'm also wondering if we can lose the `??` checks, given the `sale_cost` should always be available when `isDiscounted` is true? (same for `introductory_offer`?) Let me know if my assumption is wrong.

I'm also wondering if we should return `null` instead of 0 as the default fallback, if there is no applicable sale associated with the product. Then we can adjust `SalePriceWithInterval` to return an empty string when no sale price is applicable?

File: client/my-sites/email/email-providers-comparison/price/price-information.tsx
Comment (wongasy) [inline client/my-sites/email/email-providers-comparison/price/price-information.tsx]: I see the opportunity to make `PriceInformation` a generic component that does not rely on checking for the email provider type, and rely on the specific price information components handle the checks (e.g. like what's already been done for https://github.com/Automattic/wp-calypso/blob/4328ae9adce0844e3817d18dab00c5b52e25cb55/client/my-sites/email/email-providers-comparison/price/price-information.tsx#L85

In case of discounts, as of now, `hasDiscount( product )` will never return `true` for Titan. And when we do get a chance to test out introductory discount for Titan, it would be handy to have this condition work for both Google and Titan.

File: client/my-sites/email/email-providers-comparison/price/price-information.tsx
Comment (wongasy) [inline client/my-sites/email/email-providers-comparison/price/price-information.tsx]: I see this condition was added to render the suggested generic text. I think we should go with either this generic line or what's in place in production `Pay only ___ today - renews at ___`, and not render different messaging for pretty much the same message we're trying to convey to user (it's just that one has more info, one with less).

Let's reuse `<DiscountPriceInformation>` and get rid of this condition.

**Review by wongasy — CHANGES_REQUESTED** [review]

File: client/my-sites/email/email-providers-comparison/price/google-workspace.tsx
Comment (olaseni) [reply client/my-sites/email/email-providers-comparison/price/google-workspace.tsx]: I feel this is a small enough change.
The hard coded 1 month free trial has lapsed and is not selling anymore. There's no point leaving that in there.

File: client/my-sites/email/email-providers-comparison/price-with-interval/index.tsx
Comment (olaseni) [reply client/my-sites/email/email-providers-comparison/price-with-interval/index.tsx:22]: > I'm also wondering if we can lose the ?? checks

I'd leave those in there because the `product` object may be undefined in between render phases. The tests for discount and the retrieval of the `sale_cost` are unrelated technically as they refer to different properties of the `product`.  Also TS syntactically requires it.

>  I'm also wondering if we should return null instead of 0 

This is moot, as the function has been simplified.

File: client/my-sites/email/email-providers-comparison/price/price-information.tsx
Comment (olaseni) [reply client/my-sites/email/email-providers-comparison/price/price-information.tsx]: Done. Hopefully I have interpreted your comments well. I initially wanted to reduce the coverage of this to GW case.

File: client/my-sites/email/email-providers-comparison/price/price-information.tsx
Comment (olaseni) [reply client/my-sites/email/email-providers-comparison/price/price-information.tsx]: We can't reuse as it is; the conditions for monthly regional discounts enables renewals at a discount for several months, so the existing messaging in `<DiscountPriceInformation/>` will not always suffice.

Are suggesting a deviation from pbLl1t-1iP-p2#comment-1253

File: client/my-sites/email/email-providers-comparison/price/google-workspace.tsx
Comment (wongasy) [reply client/my-sites/email/email-providers-comparison/price/google-workspace.tsx]: @olaseni in that case, do you mind updating the test plan on how this can be tested?

File: client/my-sites/email/email-providers-comparison/price/price-information.tsx
Comment (wongasy) [reply client/my-sites/email/email-providers-comparison/price/price-information.tsx]: As discussed earlier, the reason for my suggestion to deviate from pbLl1t-1iP-p2#comment-1253, was mainly to take into account the current sale that's running, so we don't change the wording in the middle of a sale. But I am in favour of the generic disclaimer. Let's adjust the logic to favour the generic disclaimer in this PR, and **hold off merging until 50% sale is over**, so we don't change the wording mid-way through the sale.

If we use the generic messaging for all discounts, we can probably simplify the logic to
```
	if ( ! isGoogleWorkspace( product ) && ! isTitanMail( product ) ) {
		return null;
	}

	if ( ! hasDiscount( product ) || isEligibleForIntroductoryOffer ) {
		return <DiscountPriceInformation product={ product } />;
	}

	if ( isEligibleForIntroductoryOfferFreeTrial ) {
		return <FreeTrialPriceInformation product={ product } />;
	}
```

where `<DiscountPriceInformation>` simply renders the generic message
```
{ translate( 'Enjoy first year subscription at the discounted price' ) }
```
plus the tool tip.

In the future, if we were to run sales with different discount intervals, we can add a configuration in `<DiscountPriceInformation>` to render the generic messaging for different discount intervals.

To go with this approach, we should also check the regional discount billing plans to make sure the info in the `<InfoPopover>` remains valid.

**Review by wongasy — CHANGES_REQUESTED** [review]

File: client/my-sites/email/email-providers-comparison/price/google-workspace.tsx
Comment (olaseni) [reply client/my-sites/email/email-providers-comparison/price/google-workspace.tsx]: After our discussion, I have resolved to retract this change into another PR

**Comment by olaseni** [comment]
The price discount information has been switched to generic messaging that will work for numerous cases. The dynamic interpretation of free trial periods have also been retracted, to be effected in a separate iteration.

File: client/my-sites/email/email-providers-comparison/price/style.scss
Comment (wongasy) [inline client/my-sites/email/email-providers-comparison/price/style.scss:7]: I believe we are only applying the outline treatment to the extra discount? 

Screen Shot 2022-09-01 at 8 11 51 AM

File: client/my-sites/email/email-providers-comparison/price/google-workspace.tsx
Comment (wongasy) [inline client/my-sites/email/email-providers-comparison/price/google-workspace.tsx]: Let's place the `! isEligibleForIntroductoryOfferFreeTrial` check first so it's consistent with the other conditions :)

**Review by wongasy — CHANGES_REQUESTED** [review]
Just two minor comments :)

Regarding the [note about InfoPopover](https://github.com/Automattic/wp-calypso/pull/67157#discussion_r959689635), would you be able to verify if the description is valid for regional discounts?

```
This discount is only available the first time you purchase a %(googleMailService)s account, any additional mailboxes purchased after that will be at the regular price.
```

I can confirm that the discount applies to the entirety of the first purchase (including multiple mailboxes), which complies to the first part of the above description. But we should verify that any subsequent purchases will indeed be purchased at regular price (that the second part of the popover description is also valid). 

Screen Shot 2022-09-01 at 8 06 06 AM

File: client/my-sites/email/email-providers-comparison/price/style.scss
Comment (wongasy) [reply client/my-sites/email/email-providers-comparison/price/style.scss:7]: @olaseni My bad, I didn't pull the latest while testing. Confirming that the badges are rendering as expected!

* One discount
Screen Shot 2022-09-01 at 9 12 38 AM

* Extra discount
Screen Shot 2022-09-01 at 9 11 24 AM

**Review by wongasy — APPROVED** [review]
LGTM!  If you could [confirm the InfoPopover details is valid](https://github.com/Automattic/wp-calypso/pull/67157#pullrequestreview-1093285831) for regional discounts, then we're good to go. But we should hold off merging this until the Google 50% sale is over to avoid changing the text mid-way through a sale.

**Comment by wongasy** [comment]
@olaseni Thanks for making the update for conditionally showing `InfoPopover`!  Confirming that it's working as expected:

* Regional Discount - `InfoPopover` not shown

Screen Shot 2022-09-02 at 8 08 34 AM

* Coupon Sale - `InfoPopover`shown

Screen Shot 2022-09-02 at 8 10 57 AM

* Regional Discount + Coupon Sale - `InfoPopover` not shown
 
Screen Shot 2022-09-02 at 8 10 27 AM

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7512077

Thank you @olaseni for including a screenshot in the description! This is really helpful for our translators.

**Comment by olaseni** [comment]
> Hi @olaseni, could you please edit the description of this PR and add a screenshot for our translators ...

Done!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7512077).

**Pull request merged**
